src/main/java/io/lettuce/core/AbstractRedisClient.java

```diff
@@ -434,55 +434,65 @@ public void shutdown(long quietPeriod, long timeout, TimeUnit timeUnit) {
-                } else {
-                    try {
-                        closeableResource.close();
-                    } catch (Exception e) {
-                        logger.debug("Exception on Close: " + e.getMessage(), e);
-                    }
-                }
-                closeableResources.remove(closeableResource);
-            }
+    private CompletableFuture<Void> closeOtherResources() {
+        List<CompletableFuture<Void>> closeFutures = new ArrayList<>();
 
-            for (Channel c : channels) {
+        while (!closeableResources.isEmpty()) {
+            Closeable closeableResource = closeableResources.iterator().next();
 
-                ChannelPipeline pipeline = c.pipeline();
+            if (closeableResource instanceof AsyncCloseable) {
 
-                ConnectionWatchdog commandHandler = pipeline.get(ConnectionWatchdog.class);
-                if (commandHandler != null) {
-                    commandHandler.setListenOnChannelInactive(false);
+                closeFutures.add(((AsyncCloseable) closeableResource).closeAsync());
+            } else {
+                try {
+                    closeableResource.close();
+                } catch (Exception e) {
+                    logger.debug("Exception on Close: " + e.getMessage(), e);
                 }
             }
+            closeableResources.remove(closeableResource);
+        }
 
-            try {
-                closeFutures.add(toCompletableFuture(channels.close()));
-            } catch (Exception e) {
-                logger.debug("Cannot close channels", e);
-            }
+        for (Channel c : channels) {
 
-            if (!sharedResources) {
-                Future<?> groupCloseFuture = clientResources.shutdown(quietPeriod, timeout, timeUnit);
-                closeFutures.add(toCompletableFuture(groupCloseFuture));
-            } else {
-                for (EventLoopGroup eventExecutors : eventLoopGroups.values()) {
-                    Future<?> groupCloseFuture = clientResources.eventLoopGroupProvider().release(eventExecutors, quietPeriod,
-                            timeout, timeUnit);
-                    closeFutures.add(toCompletableFuture(groupCloseFuture));
-                }
+            ChannelPipeline pipeline = c.pipeline();
+
+            ConnectionWatchdog commandHandler = pipeline.get(ConnectionWatchdog.class);
+            if (commandHandler != null) {
+                commandHandler.setListenOnChannelInactive(false);
             }
+        }
 
-            return Futures.allOf(closeFutures);
+        try {
+            closeFutures.add(toCompletableFuture(channels.close()));
+        } catch (Exception e) {
+            logger.debug("Cannot close channels", e);
         }
 
-        return completedFuture(null);
+        return Futures.allOf(closeFutures);
+    }
+
+    private CompletionStage<Void> closeClientResources(long quietPeriod, long timeout, TimeUnit timeUnit) {
```

Comment:
should this return `CompletableFuture` instead of `CompletionStage` just for consistency with other methods?